src/NHibernate/Type/ComponentType.cs

```diff
@@ -520,6 +520,8 @@ public override object SemiResolve(object value, ISessionImplementor session, ob
 		{
```

Comment:
Does it fail on `AbstractEntityPersister.CacheByUniqueKeys`?

Reply:
Yes.

Full error:

```
NHibernate.Exceptions.GenericADOException : could not load an entity: [NHibernate.Test.NHSpecificTest.GH1918.BiEmbIdRefIngEntity#component[X,Y]{'X'='5', 'Y'='6'}][SQL: SELECT biembidref0_.X as x1_1_0_, biembidref0_.Y as y2_1_0_, biembidref0_.Data as data3_1_0_, biembidref0_.RefX as refx4_1_0_, biembidref0_.RefY as refy5_1_0_ FROM BiEmbIdRefIngEntity biembidref0_ WHERE biembidref0_.X=? and biembidref0_.Y=?]
  ----> System.InvalidCastException : Impossible d'effectuer un cast d'un objet de type 'NHibernate.Test.NHSpecificTest.GH1918.EmbId' en type 'System.Object[]'.
   à NHibernate.Loader.Loader.LoadEntity(ISessionImplementor session, Object id, IType identifierType, Object optionalObject, String optionalEntityName, Object optionalIdentifier, IEntityPersister persister)
   à NHibernate.Loader.Entity.AbstractEntityLoader.Load(ISessionImplementor session, Object id, Object optionalObject, Object optionalId)
   à NHibernate.Loader.Entity.AbstractEntityLoader.Load(Object id, Object optionalObject, ISessionImplementor session)
   à NHibernate.Persister.Entity.AbstractEntityPersister.Load(Object id, Object optionalObject, LockMode lockMode, ISessionImplementor session)
   à NHibernate.Event.Default.DefaultLoadEventListener.LoadFromDatasource(LoadEvent event, IEntityPersister persister, EntityKey keyToLoad, LoadType options)
   à NHibernate.Event.Default.DefaultLoadEventListener.DoLoad(LoadEvent event, IEntityPersister persister, EntityKey keyToLoad, LoadType options)
   à NHibernate.Event.Default.DefaultLoadEventListener.Load(LoadEvent event, IEntityPersister persister, EntityKey keyToLoad, LoadType options)
   à NHibernate.Event.Default.DefaultLoadEventListener.ProxyOrLoad(LoadEvent event, IEntityPersister persister, EntityKey keyToLoad, LoadType options)
   à NHibernate.Event.Default.DefaultLoadEventListener.OnLoad(LoadEvent event, LoadType loadType)
   à NHibernate.Impl.SessionImpl.FireLoad(LoadEvent event, LoadType loadType)
   à NHibernate.Impl.SessionImpl.Get(String entityName, Object id)
   à NHibernate.Impl.SessionImpl.Get(Type entityClass, Object id)
   à NHibernate.Impl.SessionImpl.Get[T](Object id)
   à NHibernate.Test.NHSpecificTest.GH1918.Fixture.CanReadBidirectionalEntitiesWithEmbeddedId() dans E:\Projets\nhibernate\nhibernate-core\src\NHibernate.Test\NHSpecificTest\GH1918\Fixture.cs:ligne 31
--InvalidCastException
   à NHibernate.Type.ComponentType.ResolveIdentifier(Object value, ISessionImplementor session, Object owner)
   à NHibernate.Type.ComponentType.SemiResolve(Object value, ISessionImplementor session, Object owner)
   à NHibernate.Persister.Entity.AbstractEntityPersister.CacheByUniqueKeys(Object entity, ISessionImplementor session)
   à NHibernate.Persister.Entity.UniqueKeyLoadableExtensions.CacheByUniqueKeys(IUniqueKeyLoadable ukLoadable, Object entity, ISessionImplementor session)
   à NHibernate.Loader.Loader.GetRow(DbDataReader rs, ILoadable[] persisters, EntityKey[] keys, Object optionalObject, EntityKey optionalObjectKey, LockMode[] lockModes, IList hydratedObjects, ISessionImplementor session, Boolean mustLoadMissingEntity)
   à NHibernate.Loader.Loader.GetRowFromResultSet(DbDataReader resultSet, ISessionImplementor session, QueryParameters queryParameters, LockMode[] lockModeArray, EntityKey optionalObjectKey, IList hydratedObjects, EntityKey[] keys, Boolean returnProxies, IResultTransformer forcedResultTransformer)
   à NHibernate.Loader.Loader.DoQuery(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer)
   à NHibernate.Loader.Loader.DoQueryAndInitializeNonLazyCollections(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies, IResultTransformer forcedResultTransformer)
   à NHibernate.Loader.Loader.DoQueryAndInitializeNonLazyCollections(ISessionImplementor session, QueryParameters queryParameters, Boolean returnProxies)
   à NHibernate.Loader.Loader.LoadEntity(ISessionImplementor session, Object id, IType identifierType, Object optionalObject, String optionalEntityName, Object optionalIdentifier, IEntityPersister persister)
```

Reply:
I think it is better to put the check there, at least it would be consistent with how other code uses `SemiResolve`.

Reply:
You are right, and my other comment, done without re-checking the code, was completely wrong.